internal/metrics/metrics.go

```diff
@@ -0,0 +1,211 @@
+	DefaultRetentionDuration = int64(7 * 24 * millisInHour)
+	// DefaultMaxBytes specifies no block data size limit
+	DefaultMaxBytes = int64(0)
+)
+
+type DataPoint struct {
+	Time  int64
+	Value float64
+}
+type Series struct {
+	Labels     map[string]string
+	DataPoints []DataPoint
+}
+
+//go:generate mockgen -package=metrics -destination=storage_mock.go . Storage
+type Storage interface {
+	io.Closer
+	Deregister() error
+	GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error)
+	AddMetric(value float64, labelsMap map[string]string) error
+}
+
+type TSDB struct {
+	appliedOptions *tsdb.Options
+	db             *tsdb.DB
+	dbLock         sync.RWMutex
+}
+
+func NewTSDB(dataDir string) (*TSDB, error) {
+	metricsDir := path.Join(dataDir, "metrics")
+	if err := os.MkdirAll(metricsDir, 0755); err != nil {
+		return nil, fmt.Errorf("cannot create directory: %w", err)
+	}
+	opts := tsdb.DefaultOptions()
+
+	opts.RetentionDuration = DefaultRetentionDuration
+	opts.MaxBytes = DefaultMaxBytes
+
+	db, err := tsdb.Open(metricsDir, nil, nil, opts, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &TSDB{
+		db:             db,
+		dbLock:         sync.RWMutex{},
+		appliedOptions: opts,
+	}, err
+}
+
+func (t *TSDB) Deregister() error {
+	metricsDir := t.db.Dir()
+	err := t.Close()
+	if err != nil {
+		return err
+	}
+
+	err = os.RemoveAll(metricsDir)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	return nil
+}
+
+func (t *TSDB) Close() error {
+	t.dbLock.Lock()
+	defer t.dbLock.Unlock()
+
+	err := t.db.Close()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	return nil
+}
+
+func (t *TSDB) GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error) {
+	log.Tracef("Getting metrics for %v - %v", tMin, tMax)
+
+	t.dbLock.RLock()
+	defer t.dbLock.RUnlock()
+
+	q, err := t.db.Querier(nil, toDbTime(tMin), toDbTime(tMax))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = q.Close()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	matchers, err := getAllLabelsMatchers(q)
+	if err != nil {
+		return nil, err
+	}
+	seriesSet := q.Select(false, nil, matchers...)
+	if seriesSet.Err() != nil {
+		log.Error(seriesSet.Err())
+		return nil, seriesSet.Err()
+	}
+
+	return toSeries(seriesSet), nil
+}
+
+func toSeries(seriesSet storage.SeriesSet) []Series {
+	var allSeries []Series
+	for {
+		if !seriesSet.Next() {
+			break
+		}
+		s := seriesSet.At()
+		labelsMap := make(map[string]string)
+		for _, label := range s.Labels() {
+			labelsMap[label.Name] = label.Value
+		}
+		var dataPoints []DataPoint
+		for i := s.Iterator(); i.Next(); {
+			t, v := i.At()
+			dataPoints = append(dataPoints, DataPoint{Time: t, Value: v})
+		}
+		allSeries = append(allSeries, Series{
+			Labels:     labelsMap,
+			DataPoints: dataPoints,
+		})
+	}
+	return allSeries
+}
+
+func (t *TSDB) AddMetric(value float64, labelsMap map[string]string) error {
+	t.dbLock.RLock()
```

Comment:
only `RLock` is needed in this context? if we modify the state of the DB, the lock for writing should be more suitable.

Reply:
Perhaps it will be better approach to use channels for submitting metrics instead of using lock to be non-blocking if that might have an impact on the workloads execution. In other words: Where can we find the caller for this method?

Reply:
There are no callers for that feature yet - we can re-think the approach, when the code that actually uses the DB is being written. Using channels looks to be a good idea.

Regarding the lock - it does not protect the content of the db, but the pointer to `TSDB` and it makes sure that the whatever it points to is visible and either completely usable or completely unusable.